src/uploadFile/pusher.ts

```diff
@@ -0,0 +1,168 @@
+): AllStatuses => {
+  if (type === 'success') {
+    return { status: Status.Success, ...data } as StatusSuccessResponse
+  }
+
+  if (type === 'progress') {
+    return { status: Status.Progress, ...data } as StatusProgressResponse
+  }
+
+  return { status: Status.Error, ...data } as StatusErrorResponse
+}
+
+const key = '79ae88bd931ea68464d9'
+const protocol =
+  typeof window !== 'undefined' && document.location.protocol !== 'https:'
+    ? 'ws:/'
+    : 'wss://'
+
+const port = protocol === 'wss://' ? 443 : 80
+
+const pusherUrl = `${protocol}ws.pusherapp.com:${port}/app/${key}?protocol=5&client=js&version=1.12.2`
+
+type Message = {
+  event: string
+  data: { channel: string }
+}
+
+type EventTypes = {
+  [key: string]: AllStatuses
+} & {
+  connected: undefined
+} & {
+  error: Error
+}
+
+class Pusher {
+  ws: WebSocket | undefined = undefined
+  queue: Message[] = []
+  isConnected = false
+  connections = 0
+  emmitter: Events<EventTypes> = new Events()
+
+  connect(): void {
+    if (!this.isConnected && !this.ws) {
+      this.ws = new WebSocket(pusherUrl)
+
+      this.ws.addEventListener('error', error => {
+        this.emmitter.emit('error', new Error(error.message))
+      })
+
+      this.emmitter.on('connected', () => {
+        this.isConnected = true
+        this.queue.forEach(message => this.send(message))
+        this.queue = []
+      })
+
+      this.ws.addEventListener('message', e => {
+        const data = JSON.parse(e.data)
+
+        switch (data.event) {
+          case 'pusher:connection_established': {
+            this.emmitter.emit('connected', undefined)
+            break
+          }
+
+          case 'progress':
+          case 'success':
+          case 'fail': {
+            this.emmitter.emit<string>(
+              data.channel,
+              response(data.event, JSON.parse(data.data))
+            )
+          }
+        }
+      })
+    }
+  }
+
+  send(data: { event: string; data: { channel: string } }): void {
+    const str = JSON.stringify(data)
+    // console.log('send: ')
+    // console.log('  ', str)
+
+    this.ws?.send(str)
+  }
+
+  subscribe(token: string, handler: (data: AllStatuses) => void): void {
+    this.connections += 1
+    this.connect()
+
+    const channel = `task-status-${token}`
+    const message = {
+      event: 'pusher:subscribe',
+      data: { channel }
+    }
+
+    this.emmitter.on(channel, handler)
+    if (this.isConnected) {
+      this.send(message)
+    } else {
+      this.queue.push(message)
+    }
+  }
+
+  unsubscribe(token: string): void {
+    this.connections -= 1
+
+    const channel = `task-status-${token}`
+    const message = {
+      event: 'pusher:unsubscribe',
+      data: { channel }
+    }
+
+    this.emmitter.off(channel)
+    if (this.isConnected) {
+      this.send(message)
+    } else {
+      this.queue = this.queue.filter(msg => msg.data.channel !== channel)
+    }
+
+    if (this.connections === 0) {
+      this.ws?.close()
```

Comment:
It would be great to implement some countdown seconds before the actual disconnect.